README.md

```diff
@@ -84,8 +84,15 @@ user to focus on the functionality, not on technical details.
 
 ## Naming things
 
-* All YAML or Python files, variables, arguments, repositories, and other such names should follow
-  standard Python naming conventions of being in snake\_case\_naming\_schemes.
+* All YAML or Python files, variables, arguments, repositories, and other such names (like
+  dictionary keys) should be valid Python identifiers following standard naming conventions of being
```

Comment:
Since the rules for valid python identifies are rather simple it would be better to spell them out like this:

```
* All YAML or Python files, variables, arguments, repositories, dictionary key and other such names should follow the following rules:
** Use only the letters A-Z (upper or lower-case), digits 0-9 or the underscore sign (`_`) as their name
** Do not use a digit as the first character
** Use snake\_case\_naming
** Rationale:
*** Jinja2 supports the dot notation (`parent.child`) to access dictionary keys. This only works when dictionary keys are valid python identifiers as specified above
*** Ansible increased similar restrictions in the past. Following this pro-actively avoids problems in the future.
```
Not sure how to best format this.

Reply:
Fine with me. Let's wait for the editors (@robled @greg-hellings) for their opinion.
By the way, you also mentioned some editor plugins as a justification for having dictionary keys as valid Python identifiers, right?

Reply:
At least my vim-yaml syntax highlighting complained about using dashes. Not sure if it would complain about starting with a digit.

Btw. I stumbled upon https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#creating-valid-variable-names which states "Variable names should be letters, numbers, and underscores. Variables should always start with a letter." but there seems to be the convention to start variables with underscores that are internal for roles...

Reply:
@tyll thanks for finding this. Starting internal variables with underscore has prior art in some popular roles - @maxamillion do you please know what is the reason for mandating that variables always start with a letter, and could the rule be changed or relaxed for role-internal variables?

Reply:
I filed https://github.com/ansible/ansible/pull/69467 to get the docs updated to allow this use case.

Reply:
Can we merge this?  At the very least, it documents our past and current guidelines/best practices, so is very valuable *right now* for developers writing new roles.  Yes, if ansible rejects https://github.com/ansible/ansible/pull/69467, or otherwise changes something in the code that makes it illegal to have variables that begin with `__`, we'll have to revise these best practices (as well as a *lot* of extant code) . . .

Reply:
I agree that we can use this now.  I saw that @greg-hellings approved the upstream PR.